### PR TITLE
typing `trainer.connectors.checkpoint connector`

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

part of #5023, finish typing for `trainer.connectors.checkpoint connector`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -430,7 +431,7 @@ def get_max_ckpt_path_from_folder(self, folder_path: Union[str, Path]) -> str:
         ckpt_number = max_suffix if max_suffix is not None else 0
         return f"{folder_path}/hpc_ckpt_{ckpt_number}.ckpt"
 
-    def save_checkpoint(self, filepath, weights_only: bool = False) -> None:
+    def save_checkpoint(self, filepath: str, weights_only: bool = False) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:434]: same

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -33,7 +34,7 @@
 
 
 class CheckpointConnector:
-    def __init__(self, trainer, resume_from_checkpoint: Optional[Union[str, Path]] = None):
+    def __init__(self, trainer: "pl.Trainer", resume_from_checkpoint: Optional[Union[str, Path]] = None) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:37]: let's use this everywhere for the path: https://github.com/PyTorchLightning/pytorch-lightning/blob/089ae9b3e82ddc31942e315294e31e48c0a899db/pytorch_lightning/utilities/types.py#L35

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by ananthsub** [comment]
fyi some overlap between this and https://github.com/PyTorchLightning/pytorch-lightning/pull/9434

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -261,14 +263,15 @@ def restore_lr_schedulers(self) -> None:
 
         # restore the lr schedulers
         lr_schedulers = self._loaded_checkpoint["lr_schedulers"]
-        for scheduler, lrs_state in zip(self.trainer.lr_schedulers, lr_schedulers):
-            scheduler["scheduler"].load_state_dict(lrs_state)
+        if self.trainer.lr_schedulers is not None:
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: [yes](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/accelerator.py#L59) good point

**Review by kaushikb11 — APPROVED** [review]

**Comment by ananthsub** [comment]
@jjenniferdai you can update this too to use the `_PATH` typehint: https://github.com/PyTorchLightning/pytorch-lightning/blob/ec828b826717cd3b5beabcb6d0cacf41b2320a98/pytorch_lightning/plugins/training_type/training_type_plugin.py#L155

File: pytorch_lightning/trainer/properties.py [hunk 4]
```diff
@@ -181,7 +181,7 @@ def optimizers(self, new_optims: Optional[List[Optimizer]]) -> None:
         self.accelerator.optimizers = new_optims
 
     @property
-    def lr_schedulers(self) -> Optional[list]:
+    def lr_schedulers(self) -> List:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/properties.py:184]: Could you update the setter below to also ensure that it only accepts a List too?

Comment (carmocca) [reply pytorch_lightning/trainer/properties.py:184]: Also, this could be annotated as `List[Union[LRSchedulerTypeTuple]]`

with `from pytorch_lightning.utilities.types import LRSchedulerTypeTuple`

Comment (ananthsub) [reply pytorch_lightning/trainer/properties.py]: > Also, this could be annotated as List[Union[LRSchedulerTypeTuple]]

this is now raising `mypy` errors. Do you think we could resolve these in a separate PR ?

Comment (jjenniferdai) [reply pytorch_lightning/trainer/properties.py]: [union directly](https://github.com/PyTorchLightning/pytorch-lightning/pull/9419/commits/b82f802e531021b5d76f27b9d80d6b97d9142730) passes if that looks good?

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 5]
```diff
@@ -689,7 +689,7 @@ def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
         checkpoint = {k: v for k, v in checkpoint.items() if k not in _exclude_keys}
         self.deepspeed_engine.save_checkpoint(filepath, client_state=checkpoint)
 
-    def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Optional[Dict[str, Any]]:
+    def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/deepspeed.py:692]:
```suggestion
    def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
```
?

File: pytorch_lightning/utilities/types.py [hunk 6]
```diff
@@ -44,4 +44,5 @@
 ]
 EVAL_DATALOADERS = Union[DataLoader, Sequence[DataLoader]]
 LRSchedulerTypeTuple = (_LRScheduler, ReduceLROnPlateau)
+LRSchedulerTypeUnion = Union[_LRScheduler, ReduceLROnPlateau]
```
Comment (ananthsub) [inline pytorch_lightning/utilities/types.py:47]: for someone who's looked at this for the first time, it's pretty confusing having `LRSchedulerTypeTuple`, `LRSchedulerTypeUnion`, and `LRSchedulerType` . i couldn't tell the difference between these based on the names. i do prefer typing the LR schedulers in a separate PR

Comment (jjenniferdai) [reply pytorch_lightning/utilities/types.py:47]: maybe rename `LRSchedulerType` --> `LRSchedulerTypeClassesUnion` ? also agree this may be better in a separate PR

Comment (ananthsub) [reply pytorch_lightning/utilities/types.py:47]: could you add a TODO here? we can merge this and then improve the LR scheduler typing separately
